operators/src/processing/expression.rs

```diff
@@ -0,0 +1,361 @@
+        source_a: Box<dyn RasterQueryProcessor<RasterType = T1>>,
+        source_b: Box<dyn RasterQueryProcessor<RasterType = T2>>,
+    ) -> Self {
+        Self {
+            source_a,
+            source_b,
+            cl_program: Self::create_cl_program(&expression),
+            phantom_data: PhantomData::default(),
+        }
+    }
+
+    fn create_cl_program(expression: &SafeExpression) -> CompiledCLProgram {
+        // TODO: generate code for arbitrary amount of inputs
+        let source = r#"
+__kernel void expressionkernel(
+            __global const IN_TYPE0 *in_data0,
+            __global const RasterInfo *in_info0,
+            __global const IN_TYPE1* in_data1,
+            __global const RasterInfo *in_info1,
+            __global OUT_TYPE0* out_data,
+            __global const RasterInfo *out_info)
+{
+    uint const gid = get_global_id(0) + get_global_id(1) * in_info0->size[0];
+    if (gid >= in_info0->size[0]*in_info0->size[1]*in_info0->size[2])
+        return;
+
+    IN_TYPE0 A = in_data0[gid];
+    if (ISNODATA0(A, in_info0)) {
+        out_data[gid] = out_info->no_data;
+        return;
+    }
+
+    IN_TYPE0 B = in_data1[gid];
+    if (ISNODATA1(B, in_info1)) {
+        out_data[gid] = out_info->no_data;
+        return;
+    }
+
+    OUT_TYPE0 result = %%%EXPRESSION%%%;
+	out_data[gid] = result;
+}"#
+        .replace("%%%EXPRESSION%%%", &expression.expression);
+
+        let mut cl_program = CLProgram::new(IterationType::Raster);
+        cl_program.add_input_raster(RasterArgument::new(T1::TYPE));
+        cl_program.add_input_raster(RasterArgument::new(T2::TYPE));
+        cl_program.add_output_raster(RasterArgument::new(TO::TYPE));
+
+        cl_program.compile(&source, "expressionkernel").unwrap()
+    }
+}
+
+impl<'a, T1, T2, TO> QueryProcessor for ExpressionQueryProcessor<T1, T2, TO>
+where
+    T1: Pixel,
+    T2: Pixel,
+    TO: Pixel,
+{
+    type Output = RasterTile2D<TO>;
+
+    fn query(
+        &self,
+        query: QueryRectangle,
+        ctx: QueryContext,
+    ) -> BoxStream<Result<RasterTile2D<TO>>> {
+        // TODO: remove if zip-map approach suffices
+        // ExpressionStream {
+        //     stream: self
+        //         .source_a
+        //         .query(query, ctx)
+        //         .zip(self.source_b.query(query, ctx)),
+        //     phantom: Default::default(),
+        // }
+        // .boxed()
+
+        // TODO: validate that tiles actually fit together, otherwise we need a different approach
+        let mut cl_program = self.cl_program.clone();
+        self.source_a
+            .query(query, ctx)
+            .zip(self.source_b.query(query, ctx))
+            .map(move |(a, b)| match (a, b) {
+                (Ok(a), Ok(b)) => {
+                    let mut out = Raster2D::new(
+                        *a.dimension(),
+                        vec![TO::zero(); a.data.data_container.len()], // TODO: correct output size; initialization required?
```

Comment:
yes. we want to initialize them. If this is really a bottleneck then we can change this later

Reply:
the kernel is executed for each kernel anway so the inizialization may be redundant but yes, we can investigate it later.